File: typescript/src/__tests__/projector.test.ts
import { Operation, type Config } from "../config";
import Projector from "../projector";


function getData() {
    return {
        projector: {
            "/": {
                "foo": "bar",
                "fem": "boy"
            },
            "/foo": {
                "foo": "bar2",
            },
            "/foo/bar": {
                "foo": "bar3",
            },

        }
    }
}

function getProjector(pwd: string, data = getData()): Projector {
    return new Projector({
        args: [],
        operation: Operation.Print,
        pwd: pwd,
        config: "Hello,something"
    }, data);
}

test("getValueAll", function() {
    const projector = getProjector("/foo/bar");

    expect(projector.getValueAll()).toEqual({
        "fem": "boy",
        "foo": "bar3"
    })
})

test("getValueKey", function() {
    const projector = getProjector("/foo/bar");
    expect(projector.getValue("foo")).toEqual("bar3")

    const projector1 = getProjector("/foo");
    expect(projector1.getValue("foo")).toEqual("bar2")
    const projector2 = getProjector("/");
    expect(projector2.getValue("foo")).toEqual("bar")
})
test("setValue", function() {
    let data = getData();
    let projector = getProjector("/foo/bar", data);
    projector.setValue("foo", "barzzz")
    expect(projector.getValue("foo")).toEqual("barzzz");
    projector = getProjector("/foo", data);
    projector.setValue("foo", "barzzz2");
    expect(projector.getValue("foo")).toEqual("barzzz2");
    projector = getProjector("/", data);
    projector.setValue("foo", "barzzz3");
    expect(projector.getValue("foo")).toEqual("barzzz3");

})

test("removeValue", function() {
    const projector = getProjector("/foo/bar");
    projector.removeValue("foo");
    expect(projector.getValue("foo")).toEqual("bar2");
})

File: go/pkg/projector/projector_test.go
package projector_test

import (
	"testing"

	"github.com/RyanDev-21/pkg/projector"
)

func getData() *projector.Data {
	return &projector.Data{
		Projector: map[string]map[string]string{
			"/": {
				"foo": "bar",
				"fem": "boy",
			},
			"/foo": {
				"foo": "bar2",
			},
			"/foo/bar": {
				"foo": "bar3",
			},
		},
	}
}

func createProjector(pwd string, data *projector.Data) *projector.Projector {
	return projector.CreateProjector(&projector.Config{
		Args:      []string{},
		Pwd:       pwd,
		Config:    "Something",
		Operation: projector.Print,
	}, data)
}

func test(t *testing.T, projector *projector.Projector, key, value string) {
	v, ok := projector.GetValue(key)
	if !ok {
		t.Error("expected to get a value")
	}
	if value != v {
		t.Errorf("expected to get %v , got %v ", value, v)
	}
}

func TestGetValue(t *testing.T) {
	data := getData()
	projector := createProjector("/foo/bar", data)
	test(t, projector, "foo", "bar3")
	projector = createProjector("/foo", data)
	test(t, projector, "foo", "bar2")
	projector = createProjector("/", data)
	test(t, projector, "foo", "bar")
}

func TestSetValue(t *testing.T) {
	data := getData()
	projector := createProjector("/foo/bar", data)
	projector.SetValue("foo", "barzzzz")
	test(t, projector, "foo", "barzzzz")
	projector.SetValue("foo", "barzzzz2")
	test(t, projector, "foo", "barzzzz2")
	projector = createProjector("/foo", data)
	projector.SetValue("foo", "barzzz2")
	test(t, projector, "foo", "barzzz2")
}

func TestRemoveValue(t *testing.T) {
	data := getData()
	projector := createProjector("/foo/bar", data)
	projector.RemoveValue("foo")
	test(t, projector, "foo", "bar2")
	projector = createProjector("/foo", data)
	projector.RemoveValue("foo")
	test(t, projector, "foo", "bar")
}
